kernel/src/grant.rs

```diff
@@ -1,266 +1,694 @@
+                // ## Explanation and Rationale
+                //
+                // This panic represents a tradeoff. While it is undesirable to
+                // have the potential for a runtime crash in this grant region
+                // code, it balances usability with type correctness. The
+                // challenge is that calling `self.apps.iter()` is a common
+                // pattern in capsules to access the grant region of every app
+                // that is using the capsule, and sometimes it is intuitive to
+                // call that inside of a `self.apps.enter(app_id, |app| {...})`
+                // closure. However, `.enter()` means that app's grant region is
+                // entered, and then a naive `.iter()` would re-enter the grant
+                // region and cause undefined behavior. We considered different
+                // options to resolve this.
+                //
+                // 1. Have `.iter()` only iterate over grant regions which are
+                //    not entered. This avoids the bug, but could lead to
+                //    unexpected behavior, as `self.apps.iter()` will do
+                //    different things depending on where in a capsule it is
+                //    called.
+                // 2. Have the compiler detect when `.iter()` is called when a
+                //    grant region has already been entered. We don't know of a
+                //    viable way to implement this.
+                // 3. Panic if `.iter()` is called when a grant is already
+                //    entered.
+                //
+                // We decided on option 3 because it balances minimizing
+                // surprises (`self.apps.iter()` will always iterate all grants)
+                // while also protecting against the bug. We expect that any
+                // code that attempts to call `self.apps.iter()` after calling
+                // `.enter()` will immediately encounter this `panic!()` and
+                // have to be refactored before any tests will be successful.
+                // Therefore, this `panic!()` should only occur at
+                // development/testing time.
+                //
+                // ## How to fix this error
+                //
+                // If you are seeing this panic, you need to refactor your
+                // capsule to not call `.iter()` or `.each()` from inside a
+                // `.enter()` closure. That is, you need to close the grant
+                // region you are currently in before trying to iterate over all
+                // grant regions.
+                panic!("Attempted to re-enter a grant region.");
             })
+            .ok();
+
+        // # `unwrap()` Safety
+        //
+        // Only `unwrap()` if some, otherwise return early.
+        let grant_ptr = if optional_grant_ptr.is_some() {
+            optional_grant_ptr.unwrap()
+        } else {
+            return None;
+        };
```

Comment:
If we wanted to get rid of the `unwrap()` here, using an if-let should be equivalent.
```suggestion
        let grant_ptr =
            if let Some(ptr) = optional_grant_ptr {
                ptr
            } else {
                return None;
            };
```